### PR TITLE
Optimize the memory utilization of Kubernetes resolver in Application Signals

### PR DESCRIPTION
**Pull request by bjrara**
# Description of the issue
The memory utilization of CloudWatch Agent is high after enabling Application Signals. In an EKS cluster of ~7k Pods, the memory usage went up to 90% for leader instance, 70% for follower instances.

From heap profiling, it was confirmed that the issue was caused by `applicationsignalsprocessor` which list/watch Pods and Services cluster-wise for IP/Service translation.

# Description of changes
This PR does the following changes to `applicationsignalsprocessor`:
1. Reduce the memory usage of list/watching.
  It is achieved by customizing transformer in informers to empty fields that might contain big objects. By applying the changes, the memory usage went down from 90% to 50% for leader, 70% to 40% for followers.
1. Remove logics of setting RemoteService to UnknownRemoteService due to IP translation failure.
  Forcing low cardinality `RemoteService` is no longer needed as we have metrics limiting in place.
1. Fix bug in Pod IP cache with `hostNetwork` as true.
  In the current implementation, if a Pod is enabled with `hostNetwork`, its IP cache will never be cleared even after the Pod is deleted.
6. Code refactoring Kubernetes resolver.

# Result
## Baseline (w/o Application Signals)
image (1)

## Before
image (2)

## After
image

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
_Describe what tests you have done._

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
8. Run `make lint`

### CODE REVIEW

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 1]
```diff
@@ -242,110 +157,107 @@ func updateHostNetworkRecords(newPod *corev1.Pod, oldPod *corev1.Pod, ipToPod *s
...
 	}
 }
 
-func onAddOrUpdatePod(newObj, oldObj interface{}, ipToPod, podToWorkloadAndNamespace, workloadAndNamespaceToLabels *sync.Map, workloadPodCount map[string]int, isAdd bool, logger *zap.Logger, deleter Deleter) {
-	pod := newObj.(*corev1.Pod)
-
+func (p *podWatcher) onAddOrUpdatePod(pod, oldPod *corev1.Pod, isAdd bool) {
 	if isAdd {
-		handlePodAdd(pod, ipToPod)
+		p.handlePodAdd(pod)
```
Comment (mxiamxia) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:216]: nice refactoring works on these verbose functions!

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 2]
```diff
@@ -523,38 +481,48 @@ func getKubernetesResolver(platformCode, clusterName string, logger *zap.Logger)
 		// jitter calls to the kubernetes api
 		jitterSleep(jitterKubernetesAPISeconds)
 
-		sharedInformerFactory := informers.NewSharedInformerFactory(clientset, 0)
+		sharedInformerFactory := informers.NewSharedInformerFactoryWithOptions(clientset, 0, informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.ResourceVersion = ""
```
Comment (mxiamxia) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:485]: Q: is `WithTweakListOptions` with `options.ResourceVersion = ""` necessary?

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 3]
```diff
@@ -523,38 +481,48 @@ func getKubernetesResolver(platformCode, clusterName string, logger *zap.Logger)
...
+		}))
 		podInformer := sharedInformerFactory.Core().V1().Pods().Informer()
+		err = podInformer.SetTransform(filterPodIPFields)
+		if err != nil {
+			logger.Error("failed to set custom transform into pod store", zap.Error(err))
+		}
```
Comment (mxiamxia) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:491]: will Informer fallback to the default caching behavior with the full Pod metadata stored if `SetTransform` failed?

**Review by mxiamxia — COMMENTED** [review]
LGTM! Left a few comments/questions but non-blocker.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 2 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:485]: This is mainly to reduce the memory spike on start.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 3 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:491]: Yes.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 2 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go]: I reverted this part of the logics as it causes inconsistency in our stateful mappings.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 4]
```diff
@@ -501,13 +390,67 @@ func (m *ServiceToWorkloadMapper) Start(stopCh chan struct{}) {
...
+		pod.Spec.EphemeralContainers = nil
+		pod.Spec.ImagePullSecrets = nil
+		pod.Spec.HostAliases = nil
+		pod.Spec.SchedulingGates = nil
+		pod.Spec.ResourceClaims = nil
+		pod.Spec.Tolerations = nil
+		pod.Spec.Affinity = nil
+
+		pod.Status.InitContainerStatuses = nil
+		pod.Status.ContainerStatuses = nil
```
Comment (mxiamxia) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:417]: please add a comment to indicate which fields will not be reset that will be used by the processor

**Review by mxiamxia — COMMENTED** [review]
🚢

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 4 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:422]: added.

**Review by mxiamxia — DISMISSED** [review]

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 5]
```diff
@@ -525,36 +477,44 @@ func getKubernetesResolver(platformCode, clusterName string, logger *zap.Logger)
 
 		sharedInformerFactory := informers.NewSharedInformerFactory(clientset, 0)
 		podInformer := sharedInformerFactory.Core().V1().Pods().Informer()
+		err = podInformer.SetTransform(filterPodIPFields)
+		if err != nil {
+			logger.Error("failed to set custom transform into pod store", zap.Error(err))
```
Comment (lisguo) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:482]: nit: change error message to something more meaningful
```suggestion
			logger.Error("failed to minimize pod object", zap.Error(err))
```

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 6]
```diff
@@ -525,36 +477,44 @@ func getKubernetesResolver(platformCode, clusterName string, logger *zap.Logger)
...
+		}
 		serviceInformer := sharedInformerFactory.Core().V1().Services().Informer()
+		err = serviceInformer.SetTransform(filterServiceIPFields)
+		if err != nil {
+			logger.Error("failed to set custom transform into service store", zap.Error(err))
```
Comment (lisguo) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:487]: similar to above
```suggestion
			logger.Error("failed to minimize service object", zap.Error(err))
```

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 7]
```diff
@@ -501,13 +390,76 @@ func (m *ServiceToWorkloadMapper) Start(stopCh chan struct{}) {
...
 		}
 	}()
 }
 
+// filterPodIPFields removes fields that could contain large objects, and retain essential
+// fields needed for IP/name translation. The following fields must be kept:
+// - ObjectMeta: Namespace, Name, Labels, OwnerReference
+// - Spec: HostNetwork, ContainerPorts
+// - Status: PodIP/s, HostIP/s
+func filterPodIPFields(obj interface{}) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:405]: It was firstly done so, but turned out creating new objects increased the memory consumption to a new level.

> If k8s api adds a new pod spec for example, then we have to update this every time.
It's true, but we don't need to update every time, only when large object is introduced.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 6 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go]: Updated.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 5 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go]: Updated.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go]: Updated to `minimizePod`

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 8]
```diff
@@ -501,13 +390,76 @@ func (m *ServiceToWorkloadMapper) Start(stopCh chan struct{}) {
...
+		}
+	}
+	return obj, nil
+}
+
+// filterServiceIPFields removes fields that could contain large objects, and retain essential
+// fields needed for IP/name translation. The following fields must be kept:
+// - ObjectMeta: Namespace, Name
+// - Spec: Selectors, ClusterIP
+func filterServiceIPFields(obj interface{}) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go]: Updated to `minimizeService`.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 9]
```diff
@@ -202,150 +114,120 @@ func onAddOrUpdateService(obj interface{}, ipToServiceAndNamespace, serviceAndNa
...
 	pod := obj.(*corev1.Pod)
-	if pod.Status.PodIP != "" {
-		deleter.DeleteWithDelay(ipToPod, pod.Status.PodIP)
-	} else if pod.Status.HostIP != "" {
+	if pod.Spec.HostNetwork && pod.Status.HostIP != "" {
+		p.logger.Debug("deleting host ip from cache", zap.String("hostNetwork", pod.Status.HostIP))
 		for _, port := range getHostNetworkPorts(pod) {
-			deleter.DeleteWithDelay(ipToPod, pod.Status.HostIP+":"+port)
+			p.deleter.DeleteWithDelay(p.ipToPod, pod.Status.HostIP+":"+port)
 		}
```
Comment (jefchien) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:210]: Same here. Can reuse `p.removeHostNetworkRecords`

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 10]
```diff
@@ -202,150 +114,120 @@ func onAddOrUpdateService(obj interface{}, ipToServiceAndNamespace, serviceAndNa
...
-			ipToPod.Store(newPod.Status.PodIP, newPod.Name)
+			p.ipToPod.Store(newPod.Status.PodIP, newPod.Name)
 		}
 	}
 }
 
-func onAddOrUpdatePod(newObj, oldObj interface{}, ipToPod, podToWorkloadAndNamespace, workloadAndNamespaceToLabels *sync.Map, workloadPodCount map[string]int, isAdd bool, logger *zap.Logger, deleter Deleter) {
-	pod := newObj.(*corev1.Pod)
-
+func (p *podWatcher) onAddOrUpdatePod(pod, oldPod *corev1.Pod, isAdd bool) {
```
Comment (jefchien) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go:179]: Couldn't we determine if it's an add based on the presence of `oldPod`?

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes_utils.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+func getWorkloadAndNamespace(pod *corev1.Pod) string {
+	var workloadAndNamespace string
+	if pod.ObjectMeta.OwnerReferences != nil {
+		for _, ownerRef := range pod.ObjectMeta.OwnerReferences {
+			if workloadAndNamespace != "" {
+				break
+			}
+
+			if ownerRef.Kind == "ReplicaSet" {
```
Comment (jefchien) [inline plugins/processors/awsapplicationsignals/internal/resolver/kubernetes_utils.go:83]: nit: The kind consts are defined in `containerinsightscommon` and can be reused.

**Review by jefchien — APPROVED** [review]

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go [hunk 12]
```diff
@@ -202,150 +114,120 @@ func onAddOrUpdateService(obj interface{}, ipToServiceAndNamespace, serviceAndNa
...
-		if oldPod.Status.PodIP != "" {
-			deleter.DeleteWithDelay(ipToPod, oldPod.Status.PodIP)
+func (p *podWatcher) handlePodUpdate(newPod *corev1.Pod, oldPod *corev1.Pod) {
+	// HostNetwork is an immutable field
+	if newPod.Spec.HostNetwork && oldPod.Status.HostIP != newPod.Status.HostIP {
+		if oldPod.Status.HostIP != "" {
+			p.logger.Debug("deleting host ip from cache", zap.String("hostNetwork", oldPod.Status.HostIP))
+			for _, port := range getHostNetworkPorts(oldPod) {
+				p.deleter.DeleteWithDelay(p.ipToPod, oldPod.Status.HostIP+":"+port)
+			}
```
[the remark this answers is not in the document]
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go]: Thanks! Updated.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 9 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go]: Updated.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go (hunk 10 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes.go]: We can. Updated.

File: plugins/processors/awsapplicationsignals/internal/resolver/kubernetes_utils.go (hunk 11 above)
Comment (bjrara) [reply plugins/processors/awsapplicationsignals/internal/resolver/kubernetes_utils.go:83]: IMHO, including `containerinsightscommon` package just for removing these one-time hardcodes alone doesn't bring a lot benefits. I can add them as consts in our package if any hardcoded string should be avoided. Rn, I'm keeping them.

**Review by lisguo — APPROVED** [review]

**Review by jefchien — APPROVED** [review]

**Pull request merged**